### PR TITLE
feat(elreal): Phase B -- rational and string construction with lazy refinement

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary

Phase B of epic #873. Implements [#875](https://github.com/stillwater-sc/universal/issues/875).

Activates the lazy refinement protocol that Phase A reserved a slot for, and adds the construction-from-source machinery: rational `elreal(p, q)` and string `elreal("1/3")` / `elreal("3.14")` / `elreal("6.022e23")`.

The acceptance criterion of #875 -- `elreal("1/3").at(1) != 0`, observably distinct from `elreal(1.0/3.0)` -- is met because the rational path computes the exact residual `p - c0 * q` via the existing EFTs (`two_prod` + `two_sum`).

## What changed

- **Generator slot promoted** from reserved-comment in Phase A to a real `mutable std::function<double(std::size_t)>` member. `at(k)` calls it on cache miss; `refine_to(precision_bits)` iterates `at(0..ceil(p/53))`.
- **Rational constructor** `elreal(long long, long long)`:
  - Component 0: `double(p) / double(q)`
  - Component 1: exact residual via EFTs, exact for `|p| < 2^53`
  - Components 2+: 0.0 with a documented Phase B limitation. Arbitrary-depth McCleeary long division is deferred to Phase E/F.
  - `q == 0` -> NaN; `p == 0` -> canonical zero.
- **String constructor** `elreal(const std::string&)`:
  - Decimal: `"3.14"`, `"-2.5"`
  - Scientific: `"6.022e23"`, `"-1.0e-10"`
  - Rational: `"1/3"`, `"-22/7"`
  - Special tokens (case-insensitive): `"nan"`, `"inf"`, `"infinity"`, with optional sign
  - Parser normalises decimal / scientific to `(p, q)` form and delegates to the rational ctor. When the normalised numerator or denominator exceeds `long long` range (typical for `6.022e23`), falls back to `double` reconstruction -- leading component faithful, exact-rational refinement unavailable.
  - Silent-failure ctor pattern matching `dfloat` and `ereal`; `parse(str, elreal&)` is the bool-returning variant for callers that want to observe errors.
- **`operator double()` fix**: starts the sum at `_components[0]` instead of `0.0` so single-component `-0.0` round-trips with its sign bit preserved (otherwise `0.0 + (-0.0) == +0.0` swallows the sign).

## Cross-construction from `dd` / `qd` / `ereal`

Marked optional in #875 and **deferred** in this PR. The include-graph cost of pulling in those types' headers from inside `elreal_impl.hpp` is non-trivial; a follow-up PR can land them targeted, once their value is clearer.

## Test plan

Four test binaries under `elastic/elreal/`:

| Binary | What it covers |
|---|---|
| `elreal_api` | Phase A canary (unchanged) |
| `elreal_native_types` | round-trip for int/float/double/subnormal/inf/nan/-0.0 |
| `elreal_rational` | `elreal(1,3).at(1) != 0`; observably distinct from `elreal(1.0/3.0)`; 1/2 and 1/4 exact in one component; -3/4 exact; 1/0 -> NaN; 0/7 -> zero |
| `elreal_string` | decimals, scientific, rationals, special tokens, parse-failure rejection |

- [x] gcc 13 (`build_elreal/`): all four PASS
- [x] clang (`build_clang_elreal/`): all four PASS

Canary excerpt:
```
elreal(1, 3):
  at(0)            = 0.333333
  at(1)            = 1.85037e-17
  at(0) + at(1)    = 0.333333
  double(1.0/3.0)  = 0.333333
elreal Phase B rational construction: PASS
```

`at(1) = 1.85e-17` -- the exact correction `1/3 - double(1/3)`, which is what the acceptance criterion of #875 asks for.

## What does *not* land in this PR

- Arithmetic operators (Phase C, #876)
- Cross-construction from `dd` / `qd` / `ereal<N>` (future targeted PR)
- Deep-depth rational refinement (`k >= 2`) via McCleeary long division (Phase E/F)

## Related

- Epic: #873
- Phase A foundation that this builds on: #874 (merged in #883)
- Phase B issue: #875

## Summary by CodeRabbit

* **New Features**
  * Rational constructor with improved residual refinement and on-demand multi-component generation.
  * String parsing for decimals, scientific notation, rationals, NaN and ±infinity, with sensible fallbacks and signed-zero preservation.
  * Operator/behavior updates to preserve negative zero and to materialize components lazily.

* **Tests**
  * New test suites covering native-type round-trips, rational construction, string parsing, edge cases (NaN, infinities, ±0), overflow/underflow, and parse-failure handling.

[Review Change Stack](https://app.coderabbit.ai/change-stack/stillwater-sc/universal/pull/884?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: include/sw/universal/number/elreal/elreal_impl.hpp [hunk 1]
```diff
@@ -287,14 +395,193 @@ class elreal {
...
+		else if (c == 'e' || c == 'E') {
+			++pos;
+			break;
+		}
+		else {
+			return false;
+		}
+		++pos;
+	}
+	if (!found_digit) return false;
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/elreal/elreal_impl.hpp:553]: Fixed in 3ead8ea1. Mantissa accumulation gets the same overflow guard as the rational parser.

File: include/sw/universal/number/elreal/elreal_impl.hpp [hunk 2]
```diff
@@ -287,14 +395,193 @@ class elreal {
...
+		for (std::size_t i = slash + 1; i < str.length(); ++i) {
+			char c = str[i];
+			if (!std::isdigit(static_cast<unsigned char>(c))) return false;
+			den = den * 10 + (c - '0');
+			found_den = true;
+		}
+		if (!found_den || den == 0) return false;
+		out = elreal(negative ? -num : num, den);
+		return true;
+	}
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/elreal/elreal_impl.hpp:514]: Fixed in 3ead8ea1. Both num and den accumulation now reject on `num * 10 + digit > LLONG_MAX` before the multiply, preventing silent wrap on 19+ digit inputs.

File: include/sw/universal/number/elreal/elreal_impl.hpp [hunk 3]
```diff
@@ -287,14 +395,193 @@ class elreal {
...
+			else q *= 10;
+		}
+	}
+	else if (q_pow < 0) {
+		for (int i = 0; i < -q_pow && !overflowed; ++i) {
+			long long pmag = p < 0 ? -p : p;
+			if (pmag > std::numeric_limits<long long>::max() / 10) overflowed = true;
+			else p *= 10;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/elreal/elreal_impl.hpp:620]: Fixed in 3ead8ea1. Replaced the magnitude computation with unsigned arithmetic so LLONG_MIN cannot trigger signed-negation UB. Defensive fix: with the mantissa overflow guard above, p cannot actually reach LLONG_MIN in this path, but a future caller of the rational ctor with a hand-rolled value could.

**Comment by Ravenwater** [comment]
**Review status after CodeRabbit's first pass** (commit 3ead8ea):

| Finding | Verdict | Action |
|---|---|---|
| `elreal_impl.hpp` rational num/den overflow | **valid** | Fixed: `num * 10 + digit > LLONG_MAX` guard before each multiply |
| `elreal_impl.hpp` decimal mantissa overflow | **valid** | Fixed: same guard pattern |
| `elreal_impl.hpp` `LLONG_MIN` UB on `-p` | **valid (defensive)** | Fixed: replaced magnitude computation with `unsigned long long`. With the mantissa overflow guard above, the UB path is now unreachable through the parser, but the unsigned form protects any future direct caller of the rational ctor. |
| `string.cpp` `"-0"` signed-zero test | **valid** | Discovered that the parser was *not* preserving the sign of `-0`, contradicting the `elreal(-0.0)` round-trip behavior already in place. Fixed in both the rational and decimal/scientific paths; three new assertions cover `"-0"`, `"-0.0"`, `"-0/7"`. |
| `native_types.cpp` `!= true` style | **valid (trivial)** | Fixed: `!std::signbit(...)` |
| `string.cpp` whitespace-wrapped tokens (`"  +inf  "`) | **declining** | The parser claims leading-whitespace-only support, matching `ereal::parse()` (`include/sw/universal/number/ereal/ereal_impl.hpp:327`). Trailing whitespace is not part of the contract on either type; adopting it asymmetrically would diverge the API surface. Left as a future enhancement if a consumer needs it on both. |

Validated under gcc 13 and clang; all four `elreal_*` binaries PASS.

File: include/sw/universal/number/elreal/elreal_impl.hpp [hunk 4]
```diff
@@ -287,14 +395,231 @@ class elreal {
...
+	long long mantissa = 0;
+	bool found_digit = false;
+	bool seen_dot = false;
+	int frac_digits = 0;
+	while (pos < str.length()) {
+		char c = str[pos];
+		if (std::isdigit(static_cast<unsigned char>(c))) {
+			int digit = c - '0';
+			if (mantissa > (std::numeric_limits<long long>::max() - digit) / 10) return false;
+			mantissa = mantissa * 10 + digit;
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/elreal/elreal_impl.hpp]: Fixed in 3016552b. Mantissa overflow now sets a flag and keeps scanning to validate the rest of the literal; after the parse loop, if the flag is set, std::stod(str) on the original input produces the correctly-rounded leading double. Tests in elastic/elreal/conversion/string.cpp gain two cases: "12345678901234567890" (20-digit integer mantissa) and "1.234567890123456789012345e5" (25-digit scientific mantissa). Both produce non-zero results matching std::stod.

File: include/sw/universal/number/elreal/elreal_impl.hpp [hunk 5]
```diff
@@ -287,14 +395,260 @@ class elreal {
...
+					// Stop updating mantissa once it would overflow, but keep
+					// scanning to validate the rest of the literal. The
+					// overflow fallback below will produce a faithful leading
+					// double via std::stod() on the full input string.
+					mantissa_overflow = true;
+				}
+				else {
+					mantissa = mantissa * 10 + digit;
+				}
+			}
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/elreal/elreal_impl.hpp:537]: Fixed in 901ec08b. Two changes:

1. Mirrored the mantissa overflow guard on the exponent loop: `exponent_overflow` set when `exp > (INT_MAX - digit) / 10`, scanning continues for format validation, fallback now routes on `mantissa_overflow || exponent_overflow`.

2. Switched the fallback from `std::stod` to `std::strtod`. `std::stod` throws `std::out_of_range` for inputs like `"1e9999999999"` (my initial attempt at this fix failed under both gcc and clang for exactly that reason). `std::strtod` sets `errno = ERANGE` and returns `+/-HUGE_VAL` or `0.0` -- IEEE-754 saturation semantics, which is the right behavior for inputs the rational track cannot represent.

Two new tests cover this:
- `elreal("1e9999999999")` must saturate to +inf
- `elreal("1e-9999999999")` must underflow to zero

All four elreal_* binaries PASS on gcc 13 and clang.
